kani-driver/src/project.rs

```diff
@@ -0,0 +1,315 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! This module defines the structure for a Kani project.
+//! The goal is to provide one project view independent on the build system (cargo / standalone
+//! rustc) and its configuration (e.g.: linker type).
+//!
+//! For `--legacy-linker` and `--function`, we still have a hack in-place that merges all the
+//! artifacts together. The reason is the following:
+//!  - For `--legacy-linker`, the compiler generate one goto file per crate, including each
+//!    target and all their dependencies. Ideally, the linking be done on a per-target base, but
+//!    this would require extra logic. Since we are planning to remove the `--legacy-linker` in
+//!    the near future, we decided not to invest time there.
+//!  - For `--function`, the compiler doesn't generate any metadata that indicates which
+//!    functions each goto model includes. Thus, we don't have an easy way to tell which goto
+//!    files are relevant for the function verification. This is also another flag that we don't
+//!    expect to stabilize, so we also opted to use the same hack of merging everything together.
+//!
+//! Note that for `--function` we also inject a mock `HarnessMetadata` to the project. This
+//! allows the rest of the driver to handle a function under verification the same way it handle
+//! other harnesses.
+
+use crate::metadata::{from_json, merge_kani_metadata, mock_proof_harness};
+use crate::session::KaniSession;
+use crate::util::{crate_name, guess_rlib_name};
+use anyhow::Result;
+use kani_metadata::{
+    artifact::convert_type, ArtifactType, ArtifactType::*, HarnessMetadata, KaniMetadata,
+};
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::BufWriter;
+use std::ops::Deref;
+use std::path::{Path, PathBuf};
+use tracing::{debug, trace};
+
+/// This structure represent the project information relevant for verification.
+/// A `Project` contains information about all crates under verification, as well as all
+/// artifacts relevant for verification.
+///
+/// For one specific harness, there should be up to one artifact of each type. I.e., artifacts of
+/// the same type are linked as part of creating the project. Except for `symtab` files used for
+/// demangling `C` when on legacy linker mode.
```

Comment:
Does this comment still apply given that this PR makes `--legacy-linker` conflict with `--gen-c`?